demos/vc_issuer/src/main.rs

```diff
@@ -0,0 +1,521 @@
+            caller(),
+            alias.id_dapp
+        )));
+    }
+    Ok(())
+}
+
+fn verify_id_alias(alias: &SignedIdAlias) -> Result<(), IssueCredentialError> {
+    let alias_tuple = AliasTuple {
+        id_alias: alias.id_alias,
+        id_dapp: alias.id_dapp,
+    };
+
+    CONFIG.with_borrow(|config| {
+        let config = config.get();
+
+        for idp_canister_id in &config.idp_canister_ids {
+            if verify_id_alias_credential_jws(
+                &alias.credential_jws,
+                &alias_tuple,
+                idp_canister_id,
+                &config.ic_root_key_raw,
+            )
+            .is_ok()
+            {
+                return Ok(());
+            }
+        }
+        Err(IssueCredentialError::InvalidIdAlias(
+            "id alias could not be verified".to_string(),
+        ))
+    })
+}
+
+#[update]
+#[candid_method]
+async fn prepare_credential(req: PrepareCredentialRequest) -> PrepareCredentialResponse {
+    if let Err(e) = authorize_vc_request(&req.signed_id_alias) {
+        return PrepareCredentialResponse::Err(e);
+    };
+    if let Err(err) = verify_credential_spec(&req.credential_spec) {
+        return PrepareCredentialResponse::Err(IssueCredentialError::UnsupportedCredentialSpec(
+            err,
+        ));
+    }
+
+    let credential = match prepare_credential_payload(&req) {
+        Ok(credential) => credential,
+        Err(err) => return PrepareCredentialResponse::Err(err),
+    };
+    let seed = calculate_seed(&req.signed_id_alias.id_alias);
+    let canister_id = ic_cdk::id();
+    let canister_sig_pk = CanisterSigPublicKey::new(canister_id, seed.to_vec());
+    let credential_jwt = credential
+        .serialize_jwt()
+        .expect("internal: JWT serialization failure");
+    let signing_input =
+        vc_signing_input(&credential_jwt, &canister_sig_pk).expect("failed getting signing_input");
+    let msg_hash = vc_signing_input_hash(&signing_input);
+
+    SIGNATURES.with(|sigs| {
+        let mut sigs = sigs.borrow_mut();
+        add_signature(&mut sigs, msg_hash, seed);
+    });
+    update_root_hash();
+    PrepareCredentialResponse::Ok(PreparedCredentialData {
+        prepared_context: Some(ByteBuf::from(credential_jwt.as_bytes())),
+    })
+}
+
+fn update_root_hash() {
+    use ic_certified_map::labeled_hash;
+    SIGNATURES.with(|sigs| {
+        let sigs = sigs.borrow();
+        let prefixed_root_hash = &labeled_hash(LABEL_SIG, &sigs.root_hash());
+        set_certified_data(&prefixed_root_hash[..]);
+    })
+}
+
+#[query]
+#[candid_method(query)]
+fn get_credential(req: GetCredentialRequest) -> GetCredentialResponse {
+    if let Err(e) = authorize_vc_request(&req.signed_id_alias) {
+        return GetCredentialResponse::Err(e);
+    };
+    if let Err(err) = verify_credential_spec(&req.credential_spec) {
+        return GetCredentialResponse::Err(IssueCredentialError::UnsupportedCredentialSpec(err));
+    }
```

Comment:
Here and elsewhere where applicable something like this might be more readable:

```rust

fn get_credential_inner(req: GetCredentialRequest) -> Result<IssuedCredentialData, IssueCredentialError> {
    authorize_vc_request(&req.signed_id_alias)?;
    verify_credential_spec(&req.credential_spec)
        .map_err(|err| { IssueCredentialError::UnsupportedCredentialSpec(err)} )?;

    ....
```

and call `get_credential_inner` from `get_credential` (just an example)

Reply:
Agreed in general, but in this example issuer the goal is to get to the point (i.e. show the flow of the API methods) asap, so I'd prefer not to add another indirection layer.